### PR TITLE
feat: Implement ast walk for Gno

### PR DESCRIPTION
**Pull request by peter7891**
Addresses https://github.com/gnolang/gno/issues/946

### CODE REVIEW

**Review by moul — COMMENTED** [review]
Hey, it looks good, but I'll make advanced review later.

Can you clarify the expected usage, please? I think you want to add a `gno inspect file.gno` in the CLI?

**Comment by peter7891** [comment]
> Hey, it looks good, but I'll make advanced review later.
> 
> Can you clarify the expected usage, please? I think you want to add a `gno inspect file.gno` in the CLI?

Did you read the comments in the issue https://github.com/gnolang/gno/issues/946

**Comment by moul** [comment]
> > Hey, it looks good, but I'll make advanced review later.
> > Can you clarify the expected usage, please? I think you want to add a `gno inspect file.gno` in the CLI?
> 
> Did you read the comments in the issue #946

Understood, thank you for the clarification.

I recommend updating the PR body to "Addresses #946" instead of "Solves #946" since this PR seems to be the initial step among several others.

Alternatively, you can switch the PR to draft mode and continue with the implementation if you prefer.

File: gnovm/pkg/gnolang/walk.go [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+	if f != nil && f(node) {
+		return f
+	}
+	return nil
+}
+
+// Inspect traverses an AST in depth-first order: It starts by calling
+// f(node); node must not be nil. If f returns true, Inspect invokes f
+// recursively for each of the non-nil children of node, followed by a
+// call of f(nil).
```
Comment (mvertes) [inline gnovm/pkg/gnolang/walk.go:419]: This comment needs to be updated to reflect what `Inspect` really does now. The behaviour is now close to https://pkg.go.dev/golang.org/x/tools/go/ast/astutil#Apply, you could take inspiration from it.

File: gnovm/pkg/gnolang/walk.go [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+func Inspect(node Node, in, out func(Node) bool) {
```
Comment (mvertes) [inline gnovm/pkg/gnolang/walk.go:420]: Using `pre` instead of `in` and `post` instead of `out` is maybe clearer.

**Review by mvertes — CHANGES_REQUESTED** [review]

File: gnovm/pkg/gnolang/walk.go (hunk 2 above)
Comment (moul) [reply gnovm/pkg/gnolang/walk.go:420]: Thinking about the Gnolang package API.

What do you prefer between:
- renaming to NodeInspect
- makint it a method of Node where the object is in: func (in *Node) Inspect…
- moving this to a new package
- staying like this?

Comment (mvertes) [reply gnovm/pkg/gnolang/walk.go:420]: Ideally `Inspect` should be a method of `Node` called `Walk`, but there may be a name collision. Maybe `WalkDepth` is ok.

Comment (peter7891) [reply gnovm/pkg/gnolang/walk.go]: > Ideally `Inspect` should be a method of `Node` called `Walk`, but there may be a name collision. Maybe `WalkDepth` is ok.

I talked to @mvertes  in slack and we came to an agreement that it should be left as a standalone function.

**Review by mvertes — APPROVED** [review]
LGTM

**Review by jaekwon — CHANGES_REQUESTED** [review]
Please take a look at this: https://github.com/gnolang/gno/issues/946#issuecomment-1631888452

Given that, we shouldn't need an additional walk method.

**Comment by peter7891** [comment]
I am closing this, for now, based on @jaekwon comment on the issue.
If we ever do decide to use this, we can revisit this branch.

**Pull request closed without merging**